File: projects/tos102us/recreate/atari/tostest.py
#!/usr/bin/env python3
"""tostest.py — run TOSTEST.PRG under a ROM and read back the conformance ledger it leaves in RAM.

    python3 tostest.py capture [--rom PATH]     one run -> out/tostest-*
    python3 tostest.py golden  [--rom PATH]     two runs -> golden/, if they agree on every field
    python3 tostest.py compare [--rom PATH]     one run -> diffed against golden/

WHAT THIS PROVES AND WHAT IT DOES NOT. The ledger is a list of answers TOS gave to fixed questions:
return values, pointers into the OS's own memory, a directory walk, a CRC of the screen after the
console driver drew on it. Two ROMs whose ledgers are identical answered every one of those
questions the same way. That is Tier 2 of the charter — the COMPOSITION being equal — and it is a
different claim from Tier 1, where each function is proved equal to the original in isolation.

EVERY FIELD OF EVERY RECORD IS REQUIRED TO MATCH, except in rows the PROGRAM marked masked (today:
Tgetdate, because Hatari seeds the emulated clock from the host). A masked row still has to be
present, with its id and name — a timing that vanished is a difference even though its number is not.

`golden` RUNS TWICE AND WRITES NOTHING IF THE TWO DISAGREE, so the reproducibility the golden depends
on is proved by the same invocation rather than asserted in a comment.
"""
import argparse
import difflib
import sys
from pathlib import Path

import hatari_rom
import ledger_run
from hatari_rom import OUT, capture_prefix, refuse, require_files

DISK = hatari_rom.HERE / "build" / "TOSTEST.ST"
PROGRAM = hatari_rom.HERE / "build" / "TOSTEST.PRG"
GOLDEN_STEM = "tostest"
# What v0/v1/v2 mean for this program — tostest/tostest.c's header is canonical for the per-call
# detail; these are the column titles.
COLUMNS = ("return", "second", "third")
TITLE = "TOSTEST conformance ledger: the answers TOS gave to a fixed set of calls"


def capture(rom, prefix, log_name):
    return ledger_run.capture(rom, DISK, ledger_run.LEDGER_KIND_TEST, prefix, COLUMNS, TITLE,
                              log_name, program=PROGRAM)


def describe(metrics):
    header, rom = metrics["header"], metrics["instrument"]["rom"]
    failed = [record for record in metrics["records"] if record["flags"] & ledger_run.FLAG_FAILED]
    return (f"  ROM {rom['version']}  sha256 {rom['sha256'][:16]}...\n"
            f"  {header['record_count']} records, status={header['status']}"
            f"{'' if header['status'] == 0 else ' (a call did not do what the program asked)'}\n"
            f"  failed rows: {', '.join(r['name'] for r in failed) if failed else 'none'}\n"
            f"  screen: {metrics['screen']['distinct_colours']} distinct colours")


def ledger_diff(golden_path, candidate_path):
    left, right = Path(golden_path).read_text(), Path(candidate_path).read_text()
    if left == right:
        return None
    return "\n".join(list(difflib.unified_diff(left.splitlines(), right.splitlines(),
                                               str(golden_path), str(candidate_path),
                                               lineterm="", n=1))[:80])


def do_capture(rom, prefix):
    metrics = capture(rom, prefix, "tostest-capture.log")
    print(f"TOSTEST under {rom} -> {prefix}-*")
    print(describe(metrics))
    print()
    print(Path(f"{prefix}-ledger.txt").read_text(), end="")
    return 0


def do_golden(rom):
    first = capture(rom, OUT / "tostest-golden1", "tostest-golden1.log")
    second = capture(rom, OUT / "tostest-golden2", "tostest-golden2.log")
    print(f"TOSTEST golden candidate from {rom}")
    print(describe(first))
    # A golden is what every later ROM is held to, so it is taken from a run that WORKED: a failed
    # call answers something, and pinning that answer makes the failure the standard.
    for index, metrics in enumerate((first, second), start=1):
        broken = ledger_run.failures(metrics["header"], metrics["records"])
        if broken:
            print(f"\nRUN {index} DID NOT PASS — no golden written:\n  " + "\n  ".join(broken))
            return 1
    if ledger_run.comparable(first["records"]) != ledger_run.comparable(second["records"]):
        print("\nTWO RUNS OF THE SAME ROM DISAGREE — no golden written:")
        print(ledger_diff(OUT / "tostest-golden1-ledger.txt", OUT / "tostest-golden2-ledger.txt"))
        return 1
    if first["screen"]["sha256"] != second["screen"]["sha256"]:
        print("\nTWO RUNS OF THE SAME ROM DREW DIFFERENT SCREENS — no golden written")
        return 1
    written = ledger_run.write_golden(OUT / "tostest-golden1", GOLDEN_STEM)
    print(f"\ntwo runs agree on every unmasked field -> golden/{GOLDEN_STEM}-* written "
          f"({len(written)} files)")
    return 0


def do_compare(rom):
    reference = ledger_run.golden_metrics(GOLDEN_STEM)
    metrics = capture(rom, OUT / "tostest-compare", "tostest-compare.log")
    moved = hatari_rom.instrument_differences(reference.get("instrument"), metrics["instrument"])
    if moved:
        refuse("this run was not taken with the golden's instrument, so any difference below would "
               "be the instrument's and not the ROM's:\n    " + "\n    ".join(moved))
    print(f"TOSTEST: {rom} against golden/")
    print(describe(metrics))
    differences = []
    if ledger_run.comparable(reference["records"]) != ledger_run.comparable(metrics["records"]):
        differences.append(("ledger", ledger_diff(hatari_rom.GOLDEN / f"{GOLDEN_STEM}-ledger.txt",
                                                  OUT / "tostest-compare-ledger.txt")))
    if reference["screen"]["sha256"] != metrics["screen"]["sha256"]:
        differences.append(("screen.png", f"sha256 {reference['screen']['sha256'][:16]} vs "
                                          f"{metrics['screen']['sha256'][:16]}"))
    if not differences:
        print("\nEVERY UNMASKED FIELD AND THE SCREEN ARE IDENTICAL.")
        return 0
    print(f"\n{len(differences)} surface(s) differ:")
    for name, text in differences:
        print(f"\n-- {name}\n{text}")
    return 1


def main():
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("mode", choices=("capture", "golden", "compare"))
    parser.add_argument("--rom", default=str(hatari_rom.ORIGINAL_ROM))
    parser.add_argument("--out", help="capture mode only: the path prefix the artefacts take")
    arguments = parser.parse_args()
    require_files((arguments.rom, "the ROM"), (DISK, "the floppy (make disks)"),
                  (PROGRAM, "the program the floppy carries (make prgs)"))
    OUT.mkdir(parents=True, exist_ok=True)
    prefix = capture_prefix(arguments.mode, arguments.out, OUT / "tostest")
    if arguments.mode == "capture":
        return do_capture(arguments.rom, prefix)
    if arguments.mode == "golden":
        return do_golden(arguments.rom)
    return do_compare(arguments.rom)


if __name__ == "__main__":
    sys.exit(main())

File: projects/tos102us/recreate/atari/tostest/tostest.c
/* tostest.c — TOSTEST.PRG, the conformance ledger.
 *
 * It calls a fixed set of TOS entry points with fixed arguments, records what each one answered
 * into the ledger at LEDGER_ADDR (../prg/ledger.h), publishes the ledger's magic and stops. The
 * host driver (../tostest.py) reads the block out of the machine and diffs it against the same
 * program's ledger under the ORIGINAL ROM. Tier 2 of the charter is that diff being empty.
 *
 * THE RECORD FIELDS, for this program: `v0` is the call's return value, `v1` and `v2` are whatever
 * second and third answer the call has (a pointer it filled in, a paired call's result, an index),
 * `blob` is up to eight bytes of a buffer it wrote, and `name` is the call. The per-call comments
 * below say which is which; nothing else needs to.
 *
 * IDS ARE STABLE AND GROUPED BY COMPONENT — 0x01xx BIOS, 0x02xx XBIOS, 0x03xx GEMDOS, and 0x04xx /
 * 0x05xx reserved for the VDI and the AES when those components land. An id is never reused for a
 * different call, so a ledger from today and one from a hundred calls hence line up row by row on
 * the ids they share. A repeated row (a directory walk) keeps ONE id and counts in `v2`.
 *
 * WHAT IS MASKED, and it is one thing: Tgetdate. Hatari seeds the emulated clock from the host, so
 * the date is a property of the day the capture was taken and not of the ROM. Everything else here
 * is required to be identical between two runs, which ../tostest.py proves before it is trusted as
 * a golden.
 *
 * THE CURSOR IS TURNED OFF BEFORE ANYTHING IS DRAWN, and that is not cosmetic: TOS's VT52 cursor
 * BLINKS off the vertical blank, so a CRC of the screen taken with it enabled is a coin flip on
 * where in the blink phase the capture lands. `console_cursor_off` (../prg/console.h) removes it.
 */
#include <stdint.h>
#include "console.h"
#include "ledger.h"
#include "st_screen.h"
#include "tosapi.h"

/* ---- the calls this ledger covers, by id ---------------------------------------------------- */
#define ID_KBSHIFT      0x0101
#define ID_DRVMAP       0x0102
#define ID_BCONOUT      0x0103
#define ID_PHYSBASE     0x0201
#define ID_LOGBASE      0x0202
#define ID_GETREZ       0x0203
#define ID_SCREEN_CRC   0x0204
#define ID_SVERSION     0x0301
#define ID_GETMPB       0x0302
#define ID_TGETDATE     0x0303
#define ID_MALLOC_MAX   0x0304
#define ID_MALLOC_BLOCK 0x0305      /* repeated: v2 is the block index */
#define ID_FSFIRST      0x0306
#define ID_DIRENTRY     0x0307      /* repeated: v2 is the entry index */

/* ---- fixed arguments ------------------------------------------------------------------------ */
#define KBSHIFT_READ_ONLY   (-1)            /* Kbshift's "report, do not set" mode */
#define MALLOC_LARGEST      (-1L)           /* Malloc's "how big is the biggest free block" */
#define MALLOC_BLOCKS       4
#define DIR_ENTRIES_MAX     16              /* a floppy root this project builds never has more */
static const char ROOT_PATTERN[] = "\\*.*";
static const char BANNER[] = "TOSTEST conformance ledger\r\n";

/* GEMDOS's memory parameter block: three pointers into the OS's own free/allocated lists. */
typedef struct { uint32_t mp_mfl, mp_mal, mp_rover; } Mpb;

/* The DTA a Fsfirst/Fsnext walk fills. Only the tail is documented data; the first 21 bytes are
 * GEMDOS's own scratch and are deliberately not recorded — they are not a conformance answer. */
typedef struct {
    uint8_t  d_reserved[21];
    uint8_t  d_attrib;
    uint16_t d_time;
    uint16_t d_date;
    uint32_t d_length;
    char     d_fname[14];
} Dta;

/* The host driver reads these two through the ledger's fields, and GEMDOS writes them by absolute
 * offset, so a padding byte GCC inserted would be a silent misread rather than a compile error. */
_Static_assert(sizeof(Dta) == 44, "the DTA is not GEMDOS's 44 bytes");
_Static_assert(sizeof(Mpb) == 12, "the MPB is not GEMDOS's three pointers");

static Dta dta;
static Mpb mpb;

/* ---- the screen ------------------------------------------------------------------------------ */
/* SCREEN_BYTES is ../st_screen.h's, the same number the ROM build paints into: a CRC taken over a
 * span that disagreed with the painter's would be a conformance failure with nothing wrong in
 * either program. */

static void console_string(const char *text)
{
    while (*text)
        Bconout(BCON_DEV_CON, (int16_t)(uint8_t)*text++);
}

/* ---- the ledger ------------------------------------------------------------------------------ */

static void record(uint16_t id, const char *name, uint32_t v0, uint32_t v1, uint32_t v2)
{
    ledger_add(id, name, 0, v0, v1, v2, 0, 0);
}

static void record_blob(uint16_t id, const char *name, uint16_t flags,
                        uint32_t v0, uint32_t v1, uint32_t v2, const void *blob, int blob_bytes)
{
    ledger_add(id, name, (uint16_t)(flags | LEDGER_FLAG_BLOB), v0, v1, v2, blob, blob_bytes);
}

static void probe_gemdos_identity(void)
{
    int32_t answer;

    record(ID_SVERSION, "Sversion", (uint32_t)Sversion(), 0, 0);

    /* Getmpb fills three pointers into the OS's memory lists; the return value is documented as
     * meaningless, so the answer IS the block. */
    answer = Getmpb(&mpb);
    record(ID_GETMPB, "Getmpb", (uint32_t)answer, mpb.mp_mfl, mpb.mp_mal);

    /* Masked: Hatari seeds the emulated clock from the host, so this is the day of the capture. */
    ledger_add(ID_TGETDATE, "Tgetdat", LEDGER_FLAG_MASKED, (uint32_t)Tgetdate(), 0, 0, 0, 0);
}

static void probe_bios(void)
{
    record(ID_KBSHIFT, "Kbshift", (uint32_t)Kbshift(KBSHIFT_READ_ONLY), 0, 0);
    record(ID_DRVMAP, "Drvmap", (uint32_t)Drvmap(), 0, 0);
}

/* Malloc's answers are addresses, and addresses are exactly what a composition test is for: they
 * say where the OS's free pool starts and how it hands blocks out. Every block is freed again, and
 * the Mfree result is recorded beside the address that produced it. */
static void probe_memory(void)
{
    int32_t blocks[MALLOC_BLOCKS];
    int block;

    record(ID_MALLOC_MAX, "MallocX", (uint32_t)Malloc(MALLOC_LARGEST), 0, 0);

    for (block = 0; block < MALLOC_BLOCKS; block++)
        blocks[block] = Malloc(MALLOC_BLOCK_BYTES);
    for (block = 0; block < MALLOC_BLOCKS; block++) {
        int32_t freed = blocks[block] ? Mfree((void *)blocks[block]) : 0;
        record(ID_MALLOC_BLOCK, "Malloc", (uint32_t)blocks[block], (uint32_t)freed, (uint32_t)block);
        if (blocks[block] == 0)
            ledger_fail();
    }
}

static void probe_directory(void)
{
    int32_t answer;
    int index;

    Fsetdta(&dta);
    answer = Fsfirst(ROOT_PATTERN, FA_NORMAL);
    record(ID_FSFIRST, "Fsfirst", (uint32_t)answer, 0, 0);
    if (answer != 0) {
        ledger_fail();
        return;
    }
    for (index = 0; index < DIR_ENTRIES_MAX && answer == 0; index++) {
        /* The disk's own file names are chosen distinct in their first LEDGER_BLOB_BYTES
         * characters, which is what the blob holds. */
        record_blob(ID_DIRENTRY, "DirEnt", 0, dta.d_length,
                    ((uint32_t)dta.d_time << 16) | dta.d_date, (uint32_t)index,
                    dta.d_fname, LEDGER_BLOB_BYTES);
        answer = Fsnext();
    }
}

/* The screen answers two ways: where the OS put it, and what is in it after the console driver has
 * been made to draw. The CRC is the drawing test — a VT52 that scrolls differently, or a font that
 * differs by one pixel, moves it. */
static void probe_screen(void)
{
    int32_t physical, logical;

    record(ID_GETREZ, "Getrez", (uint32_t)Getrez(), 0, 0);
    physical = Physbase();
    logical = Logbase();
    record(ID_PHYSBASE, "Physbas", (uint32_t)physical, 0, 0);
    record(ID_LOGBASE, "Logbase", (uint32_t)logical, 0, 0);

    console_string(BANNER);
    record(ID_BCONOUT, "Bconout", (uint32_t)(sizeof BANNER - 1), 0, 0);
    record(ID_SCREEN_CRC, "ScrnCRC", ledger_crc32((const void *)physical, SCREEN_BYTES),
           SCREEN_BYTES, 0);
}

int prg_main(void)
{
    console_cursor_off();

    ledger_begin(LEDGER_KIND_TEST);
    probe_gemdos_identity();
    probe_bios();
    probe_memory();
    probe_directory();
    probe_screen();
    ledger_publish();

    /* THE PROGRAM DOES NOT TERMINATE, and that is the design. The host stops the machine on the
     * ledger's magic and reads the block out of a machine that is still standing exactly where the
     * last record was written; a Pterm would hand control to the desktop, which would redraw the
     * screen the CRC above just measured and reuse the memory the ledger sits in. `_start` still
     * carries the Pterm for the day a mode wants it. */
    for (;;)
        ;
}
